File: packages/gittensory-mcp/scripts/gittensor-score-preview.py
#!/usr/bin/env python3
"""Reference local scorer for GITTENSOR_SCORE_PREVIEW_CMD.

Reads branch metadata JSON from stdin, reads changed files from repoRoot on disk,
and emits normalized score preview JSON to stdout. Requires a local entrius/gittensor checkout.
"""
from __future__ import annotations

import json
import os
import re
import sys
from pathlib import Path

# Every rule is an END/SEGMENT-anchored regex, a faithful mirror of the server isTestPath
# (src/signals/test-evidence.ts). Plain substring tokens over-matched: "/__snapshots__/" missed a root-level
# dir, and ".test.mjs" matched non-tests like `dist/widget.test.mjs.map` where the extension is not end-of-path.
_TEST_PATH_RES = (
    re.compile(r"(?:^|/)(?:tests?|spec|__tests__|__snapshots__|src/test)/", re.IGNORECASE),  # dir conventions
    re.compile(r"(?:^|/)[^/]+_test\.(?:go|py|rb|dart)$", re.IGNORECASE),  # go/py/rb/dart *_test suffix
    re.compile(r"(?:^|/)test_[^/]*\.py$", re.IGNORECASE),  # pytest test_*.py prefix
    re.compile(r"(?:^|/)[^/]+_spec\.rb$", re.IGNORECASE),  # RSpec *_spec.rb suffix
    re.compile(r"\.(?:test|spec)\.(?:ts|tsx|mts|cts|js|jsx|mjs|cjs|py|rb|rs)$", re.IGNORECASE),  # .test/.spec.<ext>
    re.compile(r"(?:^|/)[^/]+\.(?:cy|e2e)\.(?:ts|tsx|mts|cts|js|jsx|mjs|cjs)$", re.IGNORECASE),  # Cypress/Playwright
    re.compile(r"(?:^|/)\w*(?:Tests?|Spec)\.(?:java|kt|kts|scala|cs|swift|groovy)$"),  # JVM/.NET/Swift (case-sensitive)
)


def is_test_file(path: str) -> bool:
    return any(rx.search(path) for rx in _TEST_PATH_RES)


def load_gittensor(gittensor_root: str):
    root = Path(gittensor_root).resolve()
    if not root.exists():
        raise RuntimeError(f"GITTENSOR_ROOT does not exist: {root}")
    sys.path.insert(0, str(root))
    from gittensor.classes import FileChange
    from gittensor.utils.github_api_tools import FileContentPair
    from gittensor.validator.utils.load_weights import load_programming_languages, load_token_config
    from gittensor.validator.utils.tree_sitter_scoring import calculate_token_score_from_file_changes

    return FileChange, FileContentPair, load_programming_languages, load_token_config, calculate_token_score_from_file_changes


def score_with_gittensor(metadata: dict) -> dict:
    gittensor_root = metadata.get("gittensorRoot") or os.environ.get("GITTENSOR_ROOT")
    if not gittensor_root:
        raise RuntimeError("Set GITTENSOR_ROOT to a local entrius/gittensor checkout.")

    repo_root = Path(metadata.get("repoRoot") or os.getcwd()).resolve()
    (
        FileChange,
        FileContentPair,
        load_programming_languages,
        load_token_config,
        calculate_token_score_from_file_changes,
    ) = load_gittensor(gittensor_root)

    file_changes = []
    file_contents = {}
    source_lines = 0
    test_token_score = 0.0
    non_code_token_score = 0.0

    for entry in metadata.get("changedFiles") or []:
        path = str(entry.get("path") or "")
        if not path:
            continue
        additions = int(entry.get("additions") or 0)
        deletions = int(entry.get("deletions") or 0)
        status = str(entry.get("status") or "modified")
        file_changes.append(
            FileChange(
                pr_number=0,
                repository_full_name=str(metadata.get("repoFullName") or "local/repo"),
                filename=path,
                changes=max(additions + deletions, 0),
                additions=additions,
                deletions=deletions,
                status=status,
                previous_filename=entry.get("previousPath"),
            )
        )
        absolute = repo_root / path
        old_content = None
        new_content = None
        if status != "added" and absolute.exists():
            try:
                old_content = absolute.read_text(encoding="utf-8")
            except OSError:
                old_content = None
        if status != "removed":
            try:
                new_content = absolute.read_text(encoding="utf-8") if absolute.exists() else ""
            except OSError:
                new_content = None
        file_contents[path] = FileContentPair(old_content=old_content, new_content=new_content)
        if is_test_file(path):
            test_token_score += float(max(additions + deletions, 0))
        elif path.endswith((".md", ".txt", ".json", ".yaml", ".yml")):
            non_code_token_score += float(max(additions + deletions, 0))
        else:
            source_lines += max(additions + deletions, 0)

    weights = load_token_config()
    programming_languages = load_programming_languages()
    result = calculate_token_score_from_file_changes(file_changes, file_contents, weights, programming_languages)

    source_token_score = 0.0
    for file_result in result.file_results:
        if file_result.is_test_file:
            test_token_score = max(test_token_score, file_result.score)
        elif file_result.scoring_method == "line-count":
            non_code_token_score += file_result.score
        else:
            source_token_score += file_result.score

    total_token_score = float(result.total_score)
    if source_token_score <= 0 and total_token_score > 0:
        source_token_score = max(total_token_score - test_token_score - non_code_token_score, 0.0)

    return {
        "sourceTokenScore": round(source_token_score, 2),
        "totalTokenScore": round(total_token_score, 2),
        "sourceLines": int(result.total_lines or source_lines),
        "testTokenScore": round(test_token_score, 2),
        "nonCodeTokenScore": round(non_code_token_score, 2),
        "activeModel": "gittensor_tree_sitter_reference",
    }


def metadata_fallback(metadata: dict) -> dict:
    source = 0
    tests = 0
    non_code = 0
    for entry in metadata.get("changedFiles") or []:
        path = str(entry.get("path") or "")
        # Match the server/JS classifiers' case-insensitive extension check (e.g. `/i` regex flag) so an
        # upper-case native source path like `Foo.C` or `Foo.H` is not silently miscounted as non-code here.
        lower_path = path.lower()
        lines = max(int(entry.get("additions") or 0) + int(entry.get("deletions") or 0), 0)
        if is_test_file(path):
            tests += lines
        elif lower_path.endswith((".ts", ".tsx", ".mts", ".cts", ".js", ".jsx", ".mjs", ".cjs", ".py", ".rb", ".rs", ".go", ".java", ".kt", ".scala", ".sql", ".cs", ".swift", ".groovy", ".php", ".cpp", ".c", ".h", ".m")):
            source += lines
        else:
            non_code += lines
    return {
        "sourceTokenScore": source,
        "totalTokenScore": source + tests + non_code,
        "sourceLines": source,
        "testTokenScore": tests,
        "nonCodeTokenScore": non_code,
        "warnings": ["Fell back to metadata line counts because gittensor scoring was unavailable."],
    }


def main() -> int:
    raw = sys.stdin.read()
    metadata = json.loads(raw) if raw.strip() else {}
    try:
        payload = score_with_gittensor(metadata)
    except Exception as error:
        payload = metadata_fallback(metadata)
        payload.setdefault("warnings", []).insert(0, str(error))
    sys.stdout.write(json.dumps(payload))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: packages/gittensory-mcp/scripts/gittensor-score-preview.mjs
#!/usr/bin/env node
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

function isTestFile(file) {
  return (
    /(^|\/)(test|tests|spec|__tests__)\//i.test(file) ||
    /(^|\/)src\/test\//i.test(file) ||
    /(^|\/)[^/]+_test\.(go|py|rb|dart)$/i.test(file) || // Dart/Flutter `foo_test.dart` co-located with source
    /(^|\/)test_[^/]*\.py$/i.test(file) || // pytest's default `test_*.py` prefix (the suffix rule above only catches `*_test.py`)
    /(^|\/)[^/]+_spec\.rb$/i.test(file) ||
    /\.(test|spec)\.(ts|tsx|mts|cts|js|jsx|mjs|cjs|py|rb|rs)$/i.test(file) ||
    /(^|\/)[^/]+\.(cy|e2e)\.(ts|tsx|mts|cts|js|jsx|mjs|cjs)$/i.test(file) ||
    // JVM/.NET/Swift PascalCase test-class suffix (case-sensitive, matching the
    // signal classifiers) so C#/Swift/Groovy tests aren't counted as source.
    /(^|\/)\w*(Tests?|Spec)\.(java|kt|kts|scala|cs|swift|groovy)$/.test(file) ||
    /(^|\/)__snapshots__\//i.test(file)
  );
}

function isCodeFile(file) {
  return /\.(ts|tsx|mts|cts|js|jsx|mjs|cjs|py|rb|rs|kt|scala|java|go|sql|cs|swift|groovy|php|cpp|c|h|m)$/i.test(file) && !isTestFile(file);
}

function lineCount(file) {
  const additions = Number(file.additions ?? 0);
  const deletions = Number(file.deletions ?? 0);
  const total = additions + deletions;
  return Number.isFinite(total) && total > 0 ? total : 1;
}

async function readStdin() {
  const chunks = [];
  for await (const chunk of process.stdin) chunks.push(chunk);
  return Buffer.concat(chunks).toString("utf8");
}

function estimateFromMetadata(metadata) {
  const changedFiles = Array.isArray(metadata.changedFiles) ? metadata.changedFiles : [];
  let sourceTokenScore = 0;
  let testTokenScore = 0;
  let nonCodeTokenScore = 0;
  let sourceLines = 0;

  for (const file of changedFiles) {
    const path = String(file.path ?? "");
    const lines = lineCount(file);
    if (isTestFile(path)) {
      testTokenScore += lines;
      continue;
    }
    if (isCodeFile(path)) {
      sourceTokenScore += lines;
      sourceLines += lines;
      continue;
    }
    nonCodeTokenScore += lines;
  }

  return {
    sourceTokenScore,
    totalTokenScore: sourceTokenScore + testTokenScore + nonCodeTokenScore,
    sourceLines,
    testTokenScore,
    nonCodeTokenScore,
    warnings: [
      "Reference scorer used metadata line counts only; point GITTENSOR_SCORE_PREVIEW_CMD at scripts/gittensor-score-preview.py with GITTENSOR_ROOT for tree-sitter scoring.",
    ],
  };
}

async function main() {
  const raw = await readStdin();
  const metadata = raw.trim() ? JSON.parse(raw) : {};
  process.stdout.write(`${JSON.stringify(estimateFromMetadata(metadata))}\n`);
}

main().catch((error) => {
  process.stderr.write(`${error instanceof Error ? error.message : "reference_scorer_failed"}\n`);
  process.exit(1);
});
